ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,543 @@
+// prefetch would process the input transaction groups by analyzing each of the transaction groups and building
+// an execution queue that would allow us to fetch all the dependencies for the input transaction groups in order
+// and output these onto a channel.
+func (p *accountPrefetcher) prefetch(ctx context.Context) {
+	defer close(p.outChan)
+	accountTasks := make(map[basics.Address]*preloaderTask)
+	resourceTasks := make(map[accountCreatableKey]*preloaderTask)
+
+	var maxTxnGroupEntries int
+	if p.consensusParams.Application {
+		// the extra two are for the sender account data, plus the application global state
+		maxTxnGroupEntries = p.consensusParams.MaxTxGroupSize * (2 + p.consensusParams.MaxAppTxnAccounts + p.consensusParams.MaxAppTxnForeignApps + p.consensusParams.MaxAppTxnForeignAssets)
+	} else {
+		// 8 is the number of resources+account used in the AssetTransferTx, which is the largest one.
+		maxTxnGroupEntries = p.consensusParams.MaxTxGroupSize * 8
+	}
+
+	tasksQueue := allocPreloaderQueue(len(p.groups), maxTxnGroupEntries)
+
+	// totalBalances counts the total number of balances over all the transaction groups
+	totalBalances := 0
+	totalResources := 0
+
+	groupsReady := make([]groupTask, len(p.groups))
+
+	// Add fee sink to the first group
+	if len(p.groups) > 0 {
+		// the feeSinkAddr is known to be non-empty
+		feeSinkPreloader := &preloaderTask{
+			address:      &p.feeSinkAddr,
+			groups:       []*groupTask{&groupsReady[0]},
+			groupIndices: []int{0},
+		}
+		groupsReady[0].balancesCount = 1
+		accountTasks[p.feeSinkAddr] = feeSinkPreloader
+		tasksQueue.enqueue(feeSinkPreloader)
+	}
+
+	// iterate over the transaction groups and add all their account addresses to the list
+	queue := &tasksQueue
+	for i := range p.groups {
+		task := &groupsReady[i]
+		for j := range p.groups[i] {
+			stxn := &p.groups[i][j]
+			switch stxn.Txn.Type {
+			case protocol.PaymentTx:
+				loadAccountsAddAccountTask(&stxn.Txn.Receiver, task, accountTasks, queue)
+				loadAccountsAddAccountTask(&stxn.Txn.CloseRemainderTo, task, accountTasks, queue)
+			case protocol.AssetConfigTx:
+				loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.ConfigAsset), basics.AssetCreatable, task, resourceTasks, queue)
+			case protocol.AssetTransferTx:
+				loadAccountsAddResourceTask(&stxn.Txn.Sender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+				if !stxn.Txn.AssetSender.IsZero() {
+					loadAccountsAddResourceTask(&stxn.Txn.AssetSender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+					loadAccountsAddAccountTask(&stxn.Txn.AssetSender, task, accountTasks, queue)
+				}
+				if !stxn.Txn.AssetReceiver.IsZero() {
+					loadAccountsAddResourceTask(&stxn.Txn.AssetReceiver, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+					loadAccountsAddAccountTask(&stxn.Txn.AssetReceiver, task, accountTasks, queue)
```

Comment:
do you need to load the receiver account? in asset.go it seems like you only need GetAssetHolding/PutAssetHolding

Reply:
I guess if `ct.AssetAmount == 0 && ct.AssetReceiver == source && !clawback` ("self-transfer of zero amount") then you are going to call Get() on the sender/AssetReceiver. but you have that case handled I think because txn.Sender's account is always loaded and txn.AssetSender's account is loaded if non-zero.